source/static_thread_pool.cpp

```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright 2019-present Facebook, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+ #include <unifex/static_thread_pool.hpp>
+
+namespace unifex
+{
+
+static_thread_pool::static_thread_pool()
+: static_thread_pool(std::thread::hardware_concurrency())
+{}
+
+static_thread_pool::static_thread_pool(std::uint32_t threadCount) {
+    assert(threadCount > 0);
+
+    threads_.reserve(threadCount);
+    try {
+        threads_.emplace_back([this] { this->run(); });
+    } catch (...) {
+        close();
+    }
+}
+
+static_thread_pool::~static_thread_pool() {
+    close();
+}
+
+void static_thread_pool::run() noexcept {
+    std::unique_lock lk{mut_};
+    while (true) {
+        while (!queue_.empty()) {
+            auto* task = queue_.pop_front();
+            lk.unlock();
+            task->execute(task);
+            lk.lock();
+        }
+
+        if (stop_) {
+            return;
+        }
+
+        cv_.wait(lk);
+    }
+}
+
+void static_thread_pool::close() noexcept {
```

Comment:
I'd rather use `stop_token` and `sender join()` for this than `close()` and `stop_`

Reply:
Ok, so to add a `request_stop()` method to the thread-pool?

Adding an asynchronous `.join()` operation on the thread-pool could prove difficult given the current APIs for joining threads are all blocking in nature. I guess we could ask all threads to shut-down and when the last thread is about to exit it's `run()` function it invokes the receiver passed to the sender.

The destructor of the thread-pool would still need to call `.join()` on the threads, however most of the threads should already have terminated by this point.

Reply:
I've added a `request_stop()` method now to at least partially support the asynchronous shutdown use-case.